src/blob/index/bptree/core.rs

```diff
@@ -376,46 +383,51 @@ where
         Ok(())
     }
 
-    async fn validate_header(&self, buf: &mut Vec<u8>, blob_size: u64) -> Result<()> {
+    async fn validate_header(&self, buf: BytesMut, blob_size: u64) -> Result<BytesMut> {
         self.validate(blob_size)?;
-        if !Self::hash_valid(&self.header, buf)? {
+        if let Some(buf) = Self::hash_valid(&self.header, buf)? {
+            Ok(buf)
+        } else {
             let param = ValidationErrorKind::IndexChecksum;
-            return Err(Error::validation(param, "header hash mismatch").into());
+            Err(Error::validation(param, "header hash mismatch").into())
         }
-        Ok(())
     }
 
-    fn hash_valid(header: &IndexHeader, buf: &mut Vec<u8>) -> Result<bool> {
+    fn hash_valid(header: &IndexHeader, mut buf: BytesMut) -> Result<Option<BytesMut>> {
```

Comment:
No need to transfer ownership of `BytesMut`. It is better to use previous approach: take `buf` as mutable reference (`buf: &mut BytesMut`) and return `Result<bool>`

Reply:
To prevent resize it is even better to receive `buf` as a mut slice: `buf: &mut [u8]`